agent/engine/docker_task_engine.go

```diff
@@ -1629,26 +1643,56 @@ func (engine *DockerTaskEngine) startContainer(task *apitask.Task, container *ap
 	// If container is a firelens container, fluent host is needed to be added to the environment variable for the task.
 	// For the supported network mode - bridge and awsvpc, the awsvpc take the host 127.0.0.1 but in bridge mode,
 	// there is a need to wait for the IP to be present before the container using the firelens can be created.
+	//
+	// For bridge-mode ServiceConnect-enabled tasks, we inject pause container for each application container
+	// including the firelens container. Therefore, when resolving the container IP, we should be checking that
+	// of the associated pause container. In such case, the firelens container has network mode "container" since it's
+	//launched into its pause container's network namespace.
 	if container.GetFirelensConfig() != nil {
-		if !task.IsNetworkModeAWSVPC() && (container.GetNetworkModeFromHostConfig() == "" || container.GetNetworkModeFromHostConfig() == apitask.BridgeNetworkMode) {
+		if !task.IsNetworkModeAWSVPC() &&
+			(container.GetNetworkModeFromHostConfig() == "" ||
+				container.GetNetworkModeFromHostConfig() == apitask.BridgeNetworkMode ||
+				(container.GetNetworkModeFromHostConfig() == "container" && task.IsServiceConnectEnabled())) {
 			_, gotContainerIP := getContainerHostIP(dockerContainerMD.NetworkSettings)
+			if task.IsServiceConnectEnabled() {
+				targetContainer, err := task.GetBridgeModePauseContainerForTaskContainer(container)
```

Comment:
* Logic blocks between line 1380-1394, line 1657-1666 and line 1675-1682 are very similar. Should we create a new private function and reuse it in these places? 
* Should we add any error log to help customers and us debug issues?
```
logger.Error("Failed to start FireLens container", logger.Fields{
    field.TaskID:    task.GetID(),
    field.Container: container.Name,
    field.Error:     err,
   ....
})
```
* Would we add new positive and negative test cases to docker_task_engine_test.go for this change?

Reply:
I was just thinking about how to make the conditional in 1652-55 simpler -- I think it wants a little comment to explain what we expect to end up calling `task.GetBridgeModePauseContainerForTaskContainer(container)`.  Or maybe a simple private isServiceConnect() function like `return container.GetNetworkModeFromHostConfig() == "container" && task.IsServiceConnectEnabled()`

Reply:
> Logic blocks between line 1380-1394, line 1657-1666 and line 1675-1682 are very similar. Should we create a new private function and reuse it in these places?

There are a couple differences that have made it a bit hard to create a single helper method. One is for application containers (other than firelens), the other is for Firelens container itself. One does not expect failure on resolving Firelens IP, while the other is fine with that and will wait for the iP to be available. 

> Should we add any error log to help customers and us debug issues?

Sure will add

> Would we add new positive and negative test cases to docker_task_engine_test.go for this change?

Sure

Reply:
> I was just thinking about how to make the conditional in 1652-55 simpler -- I think it wants a little comment to explain what we expect to end up calling `task.GetBridgeModePauseContainerForTaskContainer(container)`. Or maybe a simple private isServiceConnect() function like `return container.GetNetworkModeFromHostConfig() == "container" && task.IsServiceConnectEnabled()`

Yeah we could do that